Clarification changes (#42)
lowRISC/style-guides: VerilogCodingStyle.md

*   When using parameterized widths, it is acceptable to simply use `1'b1` (e.g.
    when incrementing) rather than contrivances such as

>     `Bus_width'(1)`.

Reviewer: @hirooih Can you update the commit to match the comment by @sjgitty? The current text is quite different and missing the contrived example.

>     `Bus_width'(1)`.

Author: @imphil 
I might be misunderstanding something.

I committed as follows;
https://github.com/lowRISC/style-guides/pull/42/commits/f724aeba453363faa883c8b39c2ed3c82105b633

Is this not enough?

Synthesizable design modules must be designed around a zero-delay simulation
methodology. All forms of `#delay`, including `#0`, are not permitted.

> See Cliff Cumming's [Verilog Nonblocking Assignments With Delays, Myths &

Reviewer 2: 👍

assignment. This process makes some signals jump registers, potentially leading
to total protonic reversal. That's bad.

> Exception: For a clock divider blocking assignment must be used not to cause race condition.

Reviewer 2: This one is new to me, can you provide a citation?

Reviewer 3: I don't really understand the problem, sorry. Both `q_nb_ng` and `q_b_ok` are incrementing in steps of 2, as I would expect.

You're right that they sample differently: `q_nb_ng` gets the value after clk and `q_b_ok` gets the one before. But that's not a race condition: I think those semantics are guaranteed. Do you think there's a valid schedule where `q_nb_ng` gets the value `0010` at the end of this clock tick?

Incidentally, you're right about the "time slot A / time slot B" thing, sorry. I was thinking about the original Verilog scheduling semantics, which is defined in terms of triggering processes. There, you'd get the processes triggered one after the other (with no time between them, giving delta cycles). The SystemVerilog scheduling gives the same result by defining a single time slot that jumps back from NBA to active.

Author: @rswarbrick,  

It seems that I could explain the difficult part of this issue to you.

> I think those semantics are guaranteed.

You are right.  I used the word "race condition" for one between clk and clk_div_nb.

q_nb_ng and q_b_ok are D-Flip-flops which capture d (the output of D-Flip-flops in  clock domain clk).  They have to capture value 0b0010 as q_b_ok does in the case above.

end
```

> #### Wildcards in case items

Reviewer 2: 👍

> Use `case` instead of `case inside` nor `casez` whenever wildcard operator

Reviewer 2: I believe this is the right interpretation, but I'm unable to scan our repository at the moment (machine issues) to confirm that we're following this guidance. Also looking to get confirmation from @eunchan or @imphil or others.

Reviewer: @sjgitty we do have one case in the Ibex tracer where we use `casez`, but that's easy to fix.

> Use `case` instead of `case inside` nor `casez` whenever wildcard operator
> behavior is not required. When wildcard behavior is needed, use `case inside`
> (or `casez` if only Verilog-2001 is supported).

Reviewer: ```suggestion
Use `case` if wildcard operator behavior is not needed. 
Use `case inside` if wildcard operator behavior is needed.
Use `casez` if wildcard operator behavior is needed and Verilog-2001 compatibility is required.
```

(The last statement slightly contradicts this style guide, which requires SystemVerilog, but I'd still keep the `casez` guidance in here for cases where we need to please a random tool.)

Please also update the example further up in the `Case Statements` section, where `casez` is currently used.

performs exact matches for undriven `X` inputs. While this does not completely
fix the problems with symmetric wildcard matching, it is harder to accidentally
produce a `Z` input than an `X` input, so this form is preferred.
`case inside` does not treat either `X` nor `Z` in the case expression as a

> wildcard, so this form is more preferred than `casez`.

Reviewer: ```suggestion
wildcard, so this form is preferred over `casez`.
```

```systemverilog
logic signed [7:0]  a;
logic               incr;
logic signed [15:0] sum1, sum2, sum3;
initial begin
  a = 8'sh80;                        // a = -128
  incr = 1'b1;

>   sum1 = a + incr;                   // bad:  sum1 = 16'h0081 ( 129)

Reviewer 2: 👍

&#x1f44d;
```systemverilog {.good}
wire [7:0] sum = a + b;  // Continuous assignment
```

> &#x1f44e;

Reviewer 2: 👍  yes thank you this should not be suggested, might have been a leftover when this contained the DV-friendly portions

Reviewer: I agree that initializations shouldn't appear as "good code example" as shown currently, so removing it is a good thing.

However, in the FPGA world, there are valid use cases for initializations (since LUTs are actually initialized this can reduce area by saving reset logic). Without going into that discussion, I'd simply drop the example and not explicitly add it as "bad" example either.

Author: How about this?
This takes care of both cases.
